builtin/readline_osh.py

```
@@ -49,46 +48,67 @@ def __exit__(self, type, value, traceback):
-        print("Executing cmd: %s" % cmd)
 
-        # TODO: add READLINE_* env vars later
-        # assert isinstance(line_buffer, str)
-        # self.mem.SetNamed(location.LName("READLINE_LINE"),
-        #                  value.Str(line_buffer),
-        #                  scope_e.GlobalOnly,
-        #                  flags=SetExport)
+        # Set READLINE_* env vars so they're available in case the command uses them
+        state.ExportGlobalString(self.mem, "READLINE_LINE", line_buffer)
+        state.ExportGlobalString(self.mem, "READLINE_POINT", str(point))
 
         # TODO: refactor out shared code from Eval, cache parse tree?
 
         cmd_val = cmd_eval.MakeBuiltinArgv([cmd])
         status = self.eval.Run(cmd_val)
 
-        return (status, line_buffer, str(point))
+        # Retrieve READLINE_* env vars and compare for changes
+        readline_line = self._get_rl_env_var('READLINE_LINE')
+        readline_point = self._get_rl_env_var('READLINE_POINT')
+        post_line_buffer = readline_line if readline_line is not None else line_buffer
+        post_point = int(
+            readline_point) if readline_point is not None else point
+
+        self.mem.Unset(location.LName('READLINE_LINE'), scope_e.GlobalOnly)
+        self.mem.Unset(location.LName('READLINE_POINT'), scope_e.GlobalOnly)
+
+        return (status, post_line_buffer, post_point)
+
+    def _get_rl_env_var(self, envvar_name):
+        # type: (str) -> Optional[str]
+        """Retrieve the value of an env var, return None if undefined"""
+
+        envvar_val = self.mem.GetValue(envvar_name, scope_e.GlobalOnly)
+        if envvar_val.tag() == value_e.Str:
+            return cast(value.Str, envvar_val).s
+        elif envvar_val.tag() == value_e.Undef:
+            return None
+        else:
+            # bash has silent weird failures if you set the readline env vars
+            # to something besides a string, but I think an exception is better
```

Yeah this is a common issue -- bash doesn't check errors very well

Oh this is a good reason to be exception safe, which I commented about ...

Is the `TypeErr` caught somewhere?  Does it end up printing a message to `stderr`?

---

Sometimes GNU readline callbacks swallow errors, which is not great

---

As near as I can tell, the equivalent bash callback, `bash_execute_unix_command`, doesn't return a meaningful status code, so I doubt we should, either.

`bash_execute_unix_command` _does_ return 1 if it can't locate a keymap (rare, I suspect), but otherwise, it always returns 0, ignoring the return code of the bound shell commands. (The only thing the command return code is used for is to change the redrawing behavior when the return code is 124...whatever that's about.)

Just to make sure, I checked with bash 4 and 5. Running `bind -x` commands has no effect on `$?`, so it's not user-visible afaict. (Maybe ble.sh has more insight into this, though.)

----

Given this, I think there's no point in raising an exception, so I'm going to keep printing out the error, but remove the `raise`.

---

OK that makes sense, generally I think printing more errors than bash makes sense, but failing hard sometimes leads to incompatibilities